tla-types/src/test/scala/at/forsyte/apalache/tla/typecomp/BuilderTest.scala

```diff
@@ -31,6 +31,157 @@ import scala.collection.immutable.SortedMap
  *   - Tests whether a TBuilderInstruction, which is supposed to construct a well-typed operator expression actually
  *     does
  *   - Tests whether all of the inputs which would have produced an ill-typed expression actually cause `build` to fail
+ *
+ * <h1>HOW TO WRITE A NEW TEST</h1>
+ *
+ * Suppose we want to add a test for TLA+ operator (binary, in this example), `Foo(x,y) == ...` which has an internal
+ * representation of `foo` and is built by the `builder.foo` method. Assume, for the purposes of this example, that Foo
+ * takes two arguments - `x: a` and `y: a -> b` - and returns a value of type `b`, for any pair of types `a` and `b`.
+ *
+ * We add a test("foo") to TestOperCategoryBuilder, where OperCategory is the class of operators foo belongs to
+ * (TlaSetOper, ApalacheOper, etc.). Our goal is to use [[checkRun]] to test `builder.foo` on inputs of varying types.
+ * To do this, we need to determine the following:
+ *
+ * <ol>
+ *
+ * <li> The number of arguments to `builder.foo` and their types. In our example, `builder.foo` takes two
+ * [[TBuilderInstruction]] arguments. Typically, we introduce a type alias, so in our case
+ * {{{type T = (TBuilderInstruction, TBuilderInstruction)}}} </li>
+ *
+ * <li> The number and types of the automatically generated parameters. In our example, `Foo` has a polymorphic
+ * signature, that takes two type parameters `(a, b)`. Typically, we introduce a type alias, so in our case
+ * {{{type TParam = (TlaType1, TlaType1)}}} </li>
+ *
+ * <li> A description of a test run, for a fixed instance of parameters. [[checkRun]] accepts two arguments: a generator
+ * (discussed later) and a `run`, which is a predicate over `TParam`. A `run` should test the builder method (e.g.
+ * `builder.foo`) for parameterized correctness (tested against all parameters produced by the generator), and should
+ * return true every time if the method is correctly implemented. You _can_ write a custom run, but the recommended way
+ * is to use one of the pre-written templates, i.e. [[runUnary]], [[runBinary]], [[runTernary]], [[runVariadic]], or
+ * [[runVariadicWithDistinguishedFirst]] (depending on the arity of the builder method). </li>
+ *
+ * </ol>
+ *
+ * Assume we want to use the [[runBinary]] template to test our `builder.foo`. Regardless of arity, each template takes
+ * four arguments:
+ *
+ * <ol>
+ *
+ * <li> The `method`. In our case `builder.foo`. [[runBinary]] requires a binary method, other templates require methods
+ * of their corresponding arities. </li>
```

Comment:
Given how long these sections are, and the fact that the code blocks are breaking out of the list items, I suspect we'd up with better formatting and clearer exposition of these were organized with titled subsections. E.g., this could be:

```suggestion
 * == 1. The `method` to test ==
 *
 * In our case, ...
```